services/api/src/api/routes/rows.py

```diff
@@ -0,0 +1,319 @@
+    pass
+
+
+class ParquetResponseEmptyError(Exception):
+    pass
+
+
+# TODO: how to invalidate the cache when the parquet branch is created or deleted?
+@lru_cache(maxsize=128)
+def get_parquet_fs(dataset: str, hf_token: Optional[str]) -> HfFileSystem:
+    """Get the parquet filesystem for a dataset.
+
+    The parquet files are stored in a separate branch of the dataset repository (see PARQUET_REVISION)
+
+    Args:
+        dataset (str): The dataset name.
+        hf_token (Optional[str]): The token to access the filesystem.
+
+    Returns:
+        HfFileSystem: The parquet filesystem.
+    """
+    return HfFileSystem(dataset, repo_type="dataset", revision=PARQUET_REVISION, token=hf_token)
+
+
+UNSUPPORTED_FEATURES_MAGIC_STRINGS = ["Image(", "Audio(", "'binary'"]
+
+RowGroupReader = partial[Any]
+
+Index = Tuple[Any, List[RowGroupReader], Features, List[str]]
+
+
+class Indexer(Protocol):
+    def __call__(self, dataset: str, config: str, split: str) -> Index:
+        ...
+
+
+def create_index(
+    config_parquet_processing_steps: List[ProcessingStep],
+    init_processing_steps: List[ProcessingStep],
+    hf_endpoint: str,
+    hf_token: Optional[str] = None,
+) -> Indexer:
+    @lru_cache(maxsize=128)
+    def index(
+        dataset: str,
+        config: str,
+        split: str,
+    ) -> Index:
+        with StepProfiler(method="rows.index", step="all"):
+            # get the list of parquet files
+            with StepProfiler(method="rows.index", step="get list of parquet files for split"):
+                try:
+                    result = get_cache_entry_from_steps(
+                        processing_steps=config_parquet_processing_steps,
+                        dataset=dataset,
+                        config=config,
+                        split=None,
+                        init_processing_steps=init_processing_steps,
+                        hf_endpoint=hf_endpoint,
+                        hf_token=hf_token,
+                    )
+                    content = result["content"]
+                except ApiCustomError as e:
+                    raise e
+                except Exception as e:
+                    raise UnexpectedError("Could not get the list of parquet files to fetch the rows from.") from e
+                    # ^ TODO: improve the error, depending on the case
+                try:
+                    sources = sorted(
+                        f"{config}/{parquet_file['filename']}"
+                        for parquet_file in content["parquet_files"]
+                        if parquet_file["split"] == split and parquet_file["config"] == config
+                    )
+                except Exception as e:
+                    raise ParquetResponseFormatError(f"Could not parse the list of parquet files: {e}") from e
+                logging.debug(
+                    f"Found {len(sources)} parquet files for dataset={dataset}, config={config}, split={split}:"
```

Comment:
Question, could we have more than one parquet file by split?

Reply:
Yes, big splits have multiple shards (about 1GB). See https://huggingface.co/docs/datasets-server/parquet#sharded-parquet-files for example.

(by the way, https://datasets-server.huggingface.co/parquet?dataset=alexandrainst/danish-wit gives an error! -> there must be a bug somewhere, and it's the example in the docs). The parquet files are here: https://huggingface.co/datasets/alexandrainst/da-wit/tree/refs%2Fconvert%2Fparquet/alexandrainst--danish-wit

Reply:
I created an issue for the other case: https://github.com/huggingface/datasets-server/issues/980